cmd/query/app/querysvc/adjuster/spaniduniquifier.go

```diff
@@ -0,0 +1,157 @@
+	resourceSpans := traces.ResourceSpans()
+	for i := 0; i < resourceSpans.Len(); i++ {
+		rs := resourceSpans.At(i)
+		scopeSpans := rs.ScopeSpans()
+		for j := 0; j < scopeSpans.Len(); j++ {
+			ss := scopeSpans.At(j)
+			spans := ss.Spans()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+				if spans, ok := d.spansByID[span.SpanID()]; ok {
+					d.spansByID[span.SpanID()] = append(spans, span)
+				} else {
+					d.spansByID[span.SpanID()] = []ptrace.Span{span}
+				}
+			}
+		}
+	}
+}
+
+func (d *spanIDDeduper) isSharedWithClientSpan(spanID pcommon.SpanID) bool {
+	spans := d.spansByID[spanID]
+	for _, span := range spans {
+		if span.Kind() == ptrace.SpanKindClient {
+			return true
+		}
+	}
+	return false
+}
+
+func (d *spanIDDeduper) uniquifyServerSpanIDs(traces ptrace.Traces) {
+	oldToNewSpanIDs := make(map[pcommon.SpanID]pcommon.SpanID)
+	resourceSpans := traces.ResourceSpans()
+	for i := 0; i < resourceSpans.Len(); i++ {
+		rs := resourceSpans.At(i)
+		scopeSpans := rs.ScopeSpans()
+		for j := 0; j < scopeSpans.Len(); j++ {
+			ss := scopeSpans.At(j)
+			spans := ss.Spans()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+				// only replace span IDs for server-side spans that share the ID with something else
+				if span.Kind() == ptrace.SpanKindServer && d.isSharedWithClientSpan(span.SpanID()) {
+					newID, err := d.makeUniqueSpanID()
+					if err != nil {
+						jptrace.AddWarning(span, err.Error())
+						continue
+					}
+					oldToNewSpanIDs[span.SpanID()] = newID
+					span.SetParentSpanID(span.SpanID()) // previously shared ID is the new parent
+					span.SetSpanID(newID)
+				}
+			}
+		}
+	}
+	d.swapParentIDs(traces, oldToNewSpanIDs)
+}
+
+// swapParentIDs corrects ParentSpanID of all spans that are children of the server
+// spans whose IDs we made unique.
+func (*spanIDDeduper) swapParentIDs(
+	traces ptrace.Traces,
+	oldToNewSpanIDs map[pcommon.SpanID]pcommon.SpanID,
+) {
+	resourceSpans := traces.ResourceSpans()
+	for i := 0; i < resourceSpans.Len(); i++ {
+		rs := resourceSpans.At(i)
+		scopeSpans := rs.ScopeSpans()
+		for j := 0; j < scopeSpans.Len(); j++ {
+			ss := scopeSpans.At(j)
+			spans := ss.Spans()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+				if parentID, ok := oldToNewSpanIDs[span.ParentSpanID()]; ok {
+					if span.SpanID() != parentID {
+						span.SetParentSpanID(parentID)
+					}
+				}
+			}
+		}
+	}
+}
+
+// makeUniqueSpanID returns a new ID that is not used in the trace,
+// or an error if such ID cannot be generated, which is unlikely,
+// given that the whole space of span IDs is 2^64.
+func (d *spanIDDeduper) makeUniqueSpanID() (pcommon.SpanID, error) {
+	id := incrementSpanID(d.maxUsedID)
+	for id != pcommon.NewSpanIDEmpty() {
+		if _, exists := d.spansByID[id]; !exists {
+			d.maxUsedID = id
+			return id, nil
+		}
+		id = incrementSpanID(id)
+	}
+	return pcommon.NewSpanIDEmpty(), errors.New(warningTooManySpans)
```

Comment:
why not define var as `error` instead of just a string?

Reply:
this was copied over from the old adjuster - I can change it if you'd like. My guess was that its because we don't actually return this error but rather just just use the message from the error to put into the warning.

Reply:
yes, but since we do return an error, we could pre-create it in a var.